native/core/src/execution/operators/filter.rs

```diff
@@ -0,0 +1,452 @@
+    analysis_boundaries: Vec<ExprBoundaries>,
+) -> Vec<ColumnStatistics> {
+    analysis_boundaries
+        .into_iter()
+        .enumerate()
+        .map(
+            |(
+                idx,
+                ExprBoundaries {
+                    interval,
+                    distinct_count,
+                    ..
+                },
+            )| {
+                let (lower, upper) = interval.into_bounds();
+                let (min_value, max_value) = if lower.eq(&upper) {
+                    (Precision::Exact(lower), Precision::Exact(upper))
+                } else {
+                    (Precision::Inexact(lower), Precision::Inexact(upper))
+                };
+                ColumnStatistics {
+                    null_count: input_column_stats[idx].null_count.clone().to_inexact(),
+                    max_value,
+                    min_value,
+                    distinct_count: distinct_count.to_inexact(),
+                }
+            },
+        )
+        .collect()
+}
+
+/// The FilterExec streams wraps the input iterator and applies the predicate expression to
+/// determine which rows to include in its output batches
+struct FilterExecStream {
+    /// Output schema, which is the same as the input schema for this operator
+    schema: SchemaRef,
+    /// The expression to filter on. This expression must evaluate to a boolean value.
+    predicate: Arc<dyn PhysicalExpr>,
+    /// The input partition to filter.
+    input: SendableRecordBatchStream,
+    /// runtime metrics recording
+    baseline_metrics: BaselineMetrics,
+}
+
+pub(crate) fn batch_filter(
+    batch: &RecordBatch,
+    predicate: &Arc<dyn PhysicalExpr>,
+) -> Result<RecordBatch> {
+    predicate
+        .evaluate(batch)
+        .and_then(|v| v.into_array(batch.num_rows()))
+        .and_then(|array| {
+            Ok(match as_boolean_array(&array) {
+                // apply filter array to record batch
+                Ok(filter_array) => comet_filter_record_batch(batch, filter_array)?,
+                Err(_) => {
+                    return internal_err!("Cannot create filter_array from non-boolean predicates");
+                }
+            })
+        })
+}
+
+// BEGIN Comet changes
+// `FilterExec` could modify input batch or return input batch without change. Instead of always
+// adding `CopyExec` on top of it, we only copy input batch for the special case.
+pub fn comet_filter_record_batch(
+    record_batch: &RecordBatch,
+    predicate: &BooleanArray,
+) -> std::result::Result<RecordBatch, ArrowError> {
+    if predicate.true_count() == record_batch.num_rows() {
+        // special case where we just make an exact copy
+        let arrays: Vec<ArrayRef> = record_batch
+            .columns()
+            .iter()
+            .map(|array| {
+                let capacity = array.len();
+                let data = array.to_data();
+                let mut mutable = MutableArrayData::new(vec![&data], false, capacity);
```

Comment:
More of a Rust question for myself to dig into: is there a reason we can't do `array.to_data().clone()` and then call `make_array` on that? It seems inefficient to allocate a resizable data structure, truncate it with `extend()` then `freeze()` it, but I am still very new to this. Thankfully there are good benchmarks in this PR for me to explore that on my own. :)

Reply:
`clone` will have a shallow copy with the buffers are still shared, but we need actually copying the data.

Reply:
I thought `copy()` would make a shallow copy, while `clone()` would copy the data?

Reply:
The buffer implementation under the arrays holds references. `clone` will just copy the references without copying data.

Reply:
So `RecordBatch` stores a `Vec` of reference counted arrays `columns: Vec<Arc<dyn Array>>`? That makes sense to me that if we `clone()` we just bump ref counts.  But the underlying `ArrayData` should be safe to `clone()` because its underlying `Vec` will deep copy its buffer.

Reply:
Ah, I see the reference counted stuff all the way down inside `Buffer`. I understand why it won't work now. Thanks!

Reply:
This is the snippet of `Buffer` implementation in arrow-rs:

```rust
pub struct Buffer {
    /// the internal byte buffer.
    data: Arc<Bytes>,
    ...
}
```


And `ArrayData`:

```rust
pub struct ArrayData {
    ...
    buffers: Vec<Buffer>,
    ...
}
```

When you clone it, `Buffer` won't copy data for you but you only get a clone of `Arc`.